core/src/main/scala/io/finch/Endpoint.scala

```diff
@@ -78,15 +78,19 @@ trait Endpoint[A] { self =>
   /**
    * Maps this endpoint to the given function `A => Future[B]`.
    */
-  final def mapAsync[B](fn: A => Future[B]): Endpoint[B] = new Endpoint[B] {
-    final def apply(input: Input): Endpoint.Result[B] = self(input) match {
-      case EndpointResult.Matched(rem, out) =>
-        EndpointResult.Matched(rem, out.flatMapF(oa => oa.traverse(a => fn(a))))
-      case _ => EndpointResult.Skipped
-    }
+  final def mapAsync[B](fn: A => Future[B]): Endpoint[B] =
+    new Endpoint[B] with (Output[A] => Future[Output[B]]) {
```

Comment:
Oh that's really interesting. Makes sense now that I think about it, just not immediately intuitive. Thanks for explaining!